### PR TITLE
ROX-15042: non stackrox namespaces in auto upgrader

### PR DESCRIPTION
**Pull request by jschnath**
### Description

After manual testing it has come to my attention that the auto upgrader will fail to upgrade sensor if sensor is not within the Stackrox namespace. While the auto upgrader does not work for helm installs, and it's difficult to install sensor based on manifests in a different namespace, it is nonetheless possible and has happened in a customer case linked in the JIRA issue for this PR.

In the upgrader there's a static namespace.Stackrox const, used in many places. I've replaced all such instances with the call to a function that determines the namespace of sensor, as was already done in sensor.go for other purposes.

### User-facing documentation

- [ ] CHANGELOG update is not needed
- [ ] Documentation is not needed

### Testing

- [ ] inspected CI results

#### Automated testing

- [ ] contributed **no automated tests**
The code in GetSensorNamespace() is already in production within sensor.go and is used to determine sensor namespace through a namespace file and ENV. No big changes, just replacements of const namespaces with calls to this function.

#### How I validated my change

I've manually confirmed how the current upgrader without these changes fails when sensor namespace is not stackrox. I've also manually tested that upgrading works correctly, but on an older version of this PR. Perhaps I should retest it before merging

### CODE REVIEW

**Comment by jschnath** [comment]
It looks like there's a problem with the snapshot namespace being determined dynamically instead of being hardcoded as 'stackrox'. In the logs I can see "panic: namespaces "ci-op-mbdk6t84" not found". Looking into how to fix this

File: pkg/sensorupgrader/namespace.go
Comment (SimonBaeumer) [inline pkg/sensorupgrader/namespace.go:16]:
```suggestion
// fails, and defaults to stackrox otherwise.
```

Comment (jschnath) [reply pkg/sensorupgrader/namespace.go:16]: Fair, I'll change it.

Comment (jschnath) [reply pkg/sensorupgrader/namespace.go:16]: Right, done

File: pkg/sensorupgrader/namespace.go
Comment (vikin91) [reply pkg/sensorupgrader/namespace.go]: > the PR is currently not in a state where I want to merge it yet

I would appreciate marking it as draft. I once found this PR and felt bad that I have not reviewed it (the review was requested from Maple). When one would start looking at it, it would be clear that some things still need to be finished, thus draft fits perfect to its state.

**Comment by jschnath** [comment]
Since the sprint is ending I propose this smaller change to the way namespaces are queried in the upgrader. I originally wanted to query both the `SAToken` namespace as well as the `POD_NAMESPACE` env var to have a higher likelihood of one of them being set with a namespace, but I could not get the gke-upgrade-tests to run with this configuration.

I've now changed the code to instead only query `POD_NAMESPACE` (as is also done [here](https://github.com/stackrox/stackrox/blob/30b73cc44f81c68a41d55d488992529130ffb32c/sensor/kubernetes/clusterhealth/updater.go#L257-L266)) and leave the `SAToken` namespace file out of my heuristic. The `SAToken` namespace will still be used in `kubernetes/sensor/sensor.go` where I found it originally and where it has been since 2019, so that the behavior of sensor can remain consistent.

For the sensor upgrader this is still a strict upgrade since namespaces other than `Stackrox` can now be used iff `POD_NAMESPACE` is set.

File: sensor/kubernetes/sensor/sensor.go
Comment (rukletsov) [inline sensor/kubernetes/sensor/sensor.go:194]: Without an explanation this is very confusing. We have an exported function `GetSensorNamespace()` that aims to be the authoritative source on where Sensor runs. However, here it is used as a fallback. Why? What are the reasons and consequences of having two different mechanisms for determining Sensor namespace?

**Review by rukletsov — CHANGES_REQUESTED** [review]

File: sensor/kubernetes/sensor/sensor.go
Comment (jschnath) [reply sensor/kubernetes/sensor/sensor.go:194]: The idea here is to remain consistent with previous behavior in sensor, see my comment above. I could also undo this change and the code would be the same, since `sensorupgrader.GetSensorNamespace()` now queries `POD_NAMESPACE` and defaults on `stackrox` if it's unset.

 I only plugged the function in there to avoid code duplication, but if this is harder to read I can also remove it and leave a little bit of duplicate code in sensor.
 
 > GetSensorNamespace() [...] aims to be the authoritative source on where Sensor runs
 
Yes, however I do not want to inadvertently break / change sensor behavior by no longer querying the `SAToken` namespace in sensor. However querying the `SAToken` namespace in the sensor upgrader led to unexpected behavior that I was not able to fully resolve, so having `GetSensorNamespace` as it is in the PR now is a compromise since the Konflux sprint ended.

Rather then keep the upgrader on a static namespace there is now some capability for supporting other namespaces.

Comment (rukletsov) [reply sensor/kubernetes/sensor/sensor.go:194]: Yeah, I understand your motives. I talk about the effect it might have on posterity. What I think we're saying now is:
> There is `GetSensorNamespace()` and we _think_ it's the right way to determine Sensor namespace. However, Malte made it differently and we don't know why but are hesitant to change his approach now.

How I would read the code snippet above if I would stumble across it:
> `GetSensorNamespace()` does not quite work here because this case is special but we won't tell you in what sense.

I see a few options:

1. add a comment here saying we probably should delete the satoken approach
2. add a param to `GetSensorNamespace()` to turn on the satoken approach; this way you keep a single interface for the single task
3. removing satoken approach everywhere.

Reading up comments in https://issues.redhat.com/browse/ROX-6142 and given we're in the beginning of the release cycle my personal preference goes for option 3.

**Comment by jschnath** [comment]
One more change I've made is to move the namespace code from `pkg/SensorUpgrader` to `pkg/pods` since it logically makes more sense there. It's not just the SensorUpgrader that queries namespaces, you can use it for any pod.

File: pkg/pods/namespace.go
Comment (rukletsov) [inline pkg/pods/namespace.go:36]: `ConsiderSATokenNamespace`?

**Review by rukletsov — APPROVED** [review]
Approving modulo wording changes. Thank you!

File: pkg/pods/namespace.go
Comment (vikin91) [inline pkg/pods/namespace.go:41]:
```suggestion
// GetPodNamespace is a heuristic to determine in which namespace a given Pod runs.
```

**Review by vikin91 — COMMENTED** [review]
I looked at your change and the code looks good!

What I am missing is the description how you tested it. I would expect that you'd run Central with the code from your branch and then:
(1) an upgrade scenario in stackrox namespace, (2) the same in a different namespace.

File: pkg/pods/namespace.go
Comment (jschnath) [reply pkg/pods/namespace.go]: Aha, I'm bookmarking that, thanks 👍

File: pkg/pods/namespace.go
Comment (jschnath) [reply pkg/pods/namespace.go]: Done

File: pkg/pods/namespace.go
Comment (jschnath) [reply pkg/pods/namespace.go]: > I didn't understand what NO_SA_TOKEN and USE_SA_TOKEN are for so I dropped them

They were the names of the enum values before I renamed them to fit our naming conventions, oops. I'll change the comments to your version.

Comment (jschnath) [reply pkg/pods/namespace.go]: Sure

File: pkg/pods/namespace.go
Comment (jschnath) [reply pkg/pods/namespace.go:41]: Right, I'll change that

**Comment by jschnath** [comment]
> What I am missing is the description how you tested it. I would expect that you'd run Central with the code from your branch and then:
(1) an upgrade scenario in stackrox namespace, (2) the same in a different namespace.

I have done both, but on an earlier iteration of the PR with slightly different code. I could test again for peace of mind.

File: sensor/upgrader/preflight/namespace.go
Comment (rukletsov) [inline sensor/upgrader/preflight/namespace.go:40]: Do you want to keep this log line or it's for tracing? Also, there's `%q` which reads simpler as `\"%s\"`.

File: sensor/upgrader/preflight/namespace.go
Comment (rukletsov) [inline sensor/upgrader/preflight/namespace.go:48]: Use `Warnf()` for readability.

File: sensor/upgrader/preflight/namespace.go
Comment (jschnath) [reply sensor/upgrader/preflight/namespace.go]: I'm still doing some manual tests to be thorough after which I'll remove this and merge the PR.

File: sensor/upgrader/preflight/namespace.go
Comment (jschnath) [reply sensor/upgrader/preflight/namespace.go:48]: Will do

**Comment by jschnath** [comment]
Okay, I've done a lot of manual testing and a few improvements here and there and am happy with how it works now. I'll be running CI and merging if it's green. Extra logs for debugging should be removed now.

File: sensor/upgrader/preflight/namespace.go
Comment (rukletsov) [inline sensor/upgrader/preflight/namespace.go:45]: Minor, but as I mentioned before, there is probably no need in `\"%s\"` constructs here (because these entities won't contain a space), which also would be consistent with how `Errorf()` below is constructed. Moreover, prefer `%q` instead of `\"%s\"` for readability.

**Pull request merged**
